src/types/basic.rs

```diff
@@ -49,41 +54,93 @@ impl TypeClass for ClassicLeaf {
 impl TypeClass for AnyLeaf {
     const TAG: TypeTag = TypeTag::Simple;
 }
-pub struct TaggedWrapper<T>(TypeTag, T);
 
-impl<T> TaggedWrapper<T> {
-    pub fn tag(&self) -> TypeTag {
-        self.0
+#[derive(Clone, PartialEq, Debug, Eq)]
+pub struct Tagged<I, T>(I, PhantomData<T>);
```

Comment:
I was sure I had a version of this without `PhantomData` but think I must have just been imagining it because indeed I can't make it work either! So ok :), yes this is right :-(